cpp/daal/include/services/service_profiler.h

```diff
@@ -0,0 +1,468 @@
+    return verbose_value;
+}
+
+inline static bool is_profiler_enabled()
+{
+    static const bool profiler_value = [] {
+        int value = daal_verbose_mode();
+        return value == 1 || value == 2 || value == 3 || value == 4 || value == 5;
+    }();
+    return profiler_value;
+}
+
+inline static bool is_analyzer_enabled()
+{
+    static const bool profiler_value = [] {
+        int value = daal_verbose_mode();
+        return value == 3 || value == 4 || value == 5;
+    }();
+    return profiler_value;
+}
+
+struct task_entry
+{
+    std::int64_t idx;
+    std::string name;
+    std::uint64_t duration;
+    std::int64_t level;
+    std::int64_t count;
+    bool threading_task;
+};
+
+struct task
+{
+    std::vector<task_entry> kernels;
+};
+
+struct task_thread
+{
+    std::set<task_entry> kernels;
+};
+
+class profiler_task
+{
+public:
+    inline profiler_task(const char * task_name, int idx) : task_name_(task_name), idx(idx) {}
+    inline profiler_task(const char * task_name, int idx, bool thread) : task_name_(task_name), idx(idx), is_thread(thread) {}
+    inline ~profiler_task();
+
+private:
+    const char * task_name_;
+    int idx;
+    bool is_thread = false;
+};
+
+class profiler
+{
+public:
+    inline profiler()
+    {
+        int verbose = daal_verbose_mode();
+        if (verbose == 1 || verbose == 4 || verbose == 5) print_header();
+    }
+
+    inline ~profiler()
+    {
+        if (is_analyzer_enabled())
+        {
+            merge_tasks();
+            const auto & tasks_info  = get_instance()->get_task();
+            std::uint64_t total_time = 0;
+            std::cerr << "Algorithm tree analyzer" << '\n';
+
+            for (size_t i = 0; i < tasks_info.kernels.size(); ++i)
+            {
+                const auto & entry = tasks_info.kernels[i];
+                if (entry.level == 0) total_time += entry.duration;
+            }
+
+            for (size_t i = 0; i < tasks_info.kernels.size(); ++i)
+            {
+                const auto & entry = tasks_info.kernels[i];
+                std::string prefix;
+                for (std::int64_t lvl = 0; lvl < entry.level; ++lvl) prefix += "|   ";
+                bool is_last = (i + 1 < tasks_info.kernels.size()) && (tasks_info.kernels[i + 1].level >= entry.level) ? false : true;
+                prefix += is_last ? "|-- " : "|-- ";
+                std::cerr << prefix << entry.name << " time: " << format_time_for_output(entry.duration) << " " << std::fixed << std::setprecision(2)
+                          << (total_time > 0 ? (double(entry.duration) / total_time) * 100 : 0.0) << "% " << entry.count << " times"
+                          << " in a " << entry.threading_task << " region" << '\n';
+            }
+            std::cerr << "|---(end)" << '\n';
+            std::cerr << "DAAL KERNEL_PROFILER: kernels total time " << format_time_for_output(total_time) << '\n';
+        }
+    }
+
+    inline static profiler_task start_task(const char * task_name)
```

Comment:
Please add comments helping to distinguish `start_task` and `start_threading_task`. In which situations one or another is used?